### ISSUE TITLE
Kernel: Cannot run GCC after moving to shared libraries

### ISSUE THREAD

**Issue by awesomekling** [SerenityOS/serenity#4417]
The GCC port no longer works inside Serenity. Other ports I've tested work, just something weird about GCC.

cc @itamar8910

**Comment by linusg** [comment]
Adding some additional info.

---

This is what I get in HackStudio when trying to build the `little` project:

![image](https://user-images.githubusercontent.com/19366641/102147526-c9e6df00-3e62-11eb-95e4-998280e73774.png)

Debug console:

```
[HackStudio(49:49)]: exec(/usr/local/bin/make): Using program interpreter /usr/lib/Loader.so
make(49): FIXME(LibC): setlocale(0, '')
[g++(52:52)]: exec(/usr/local/lib/gcc/i686-pc-serenity/10.2.0/../../../../i686-pc-serenity/bin/as): Using program interpreter /usr/lib/Loader.so
USERSPACE(52) ASSERTION FAILED: false
../Libraries/LibELF/DynamicLoader.cpp:376
[as(52:52)]: Terminating as(52) due to signal 6
```

The code of that `ASSERT_NOT_REACHED` is (had to `dbgln()` it, building with `DYNAMIC_LOAD_VERBOSE` doesn't work):

https://github.com/SerenityOS/serenity/blob/efe4da57dfc4d820062455f41fb5f16842268d40/Libraries%2FLibELF%2FDynamicLoader.cpp#L373-L377

`relocation.type()` is `5` in this case, i.e. `R_386_COPY`:

https://github.com/SerenityOS/serenity/blob/07b49573612628cc6388e9fe30083dc9912b6a7b/Libraries%2FLibELF%2Fexec_elf.h#L785

**Comment by ADKaster** [comment]
It's surprising that we are seeing an R_386_COPY relocation.

From the linked oracle doc:
> Copy relocations can be eliminated by building the application from position-independent code. See "Position-Independent Code".

Also
> R_386_COPY
> Created by the link-editor for dynamic executables to preserve a read-only text segment. Its offset member refers to a location in a writable segment. The symbol table index specifies a symbol that should exist both in the current object file and in a shared object. During execution, the runtime linker copies data associated with the shared object's symbol to the location specified by the offset. See "Copy Relocations".

That suggests that something the gcc port link stage is using is not properly PIC code. Whether it's choosing the static libc or something? Or we are missing the proper configure flags in the ports dir, I'm not sure, haven't looked into it. Could use `ldd` or `readelf -R` to double check the compiled libs that gcc is trying to load to see which ones (or maybe the executables themselves? cc1?) has the suspect relocations generated.

**Comment by itamar8910** [comment]
Output of `readelf -h gcc`:
```
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF32
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           Intel 80386
  Version:                           0x1
  Entry point address:               0x804e441
  Start of program headers:          52 (bytes into file)
  Start of section headers:          1153124 (bytes into file)
  Flags:                             0x0
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         5
  Size of section headers:           40 (bytes)
  Number of section headers:         28
  Section header string table index: 27
```
Output of `readelf -d gcc`:
```
Dynamic section at offset 0xfb1f4 contains 21 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so]
 0x00000001 (NEEDED)                     Shared library: [libm.so]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
 0x0000000c (INIT)                       0x80496fc
 0x0000000d (FINI)                       0x80d6a51
 0x00000019 (INIT_ARRAY)                 0x8120000
 0x0000001b (INIT_ARRAYSZ)               20 (bytes)
 0x00000004 (HASH)                       0x80480e8
 0x00000005 (STRTAB)                     0x8048e14
 0x00000006 (SYMTAB)                     0x8048534
 0x0000000a (STRSZ)                      1206 (bytes)
 0x0000000b (SYMENT)                     16 (bytes)
 0x00000015 (DEBUG)                      0x0
 0x00000003 (PLTGOT)                     0x8143344
 0x00000002 (PLTRELSZ)                   992 (bytes)
 0x00000014 (PLTREL)                     REL
 0x00000017 (JMPREL)                     0x804931c
 0x00000011 (REL)                        0x80492cc
 0x00000012 (RELSZ)                      80 (bytes)
 0x00000013 (RELENT)                     8 (bytes)
 0x00000000 (NULL)                       0x0
```
So gcc does use shared libs, but it's just not a PIE, and the problem is that we currently don't support loading non-PIE programs.
I'm working on fixing this, will update.

**Comment by itamar8910** [comment]
I have a [branch](https://github.com/itamar8910/serenity/tree/fix-gcc) on my fork that fixes gcc, but as a side effect it breaks static executables (we currently don't build anything statically, though). 
A more proper fix requires us to do less hacky things with our `crt0` - will update on that.

**Comment by itamar8910** [comment]
fixed in #4503
